### DISCUSSION TITLE
Add event before preparing an email to send out

### DISCUSSION THREAD

**Discussion by modelesque** [craftcms/commerce#2983] [Ideas] [answered]
I wish there was an event that would allow me to adjust the contents of an `Email` model before it gets prepared in `\craft\commerce\services\Emails::sendEmail` and sent out. 

One issue that I'm facing is that our emails don't just go to the order's customer or to our company's email address. Sometimes we need to send a third type of email to a set of users in a `myRelatedUsers` field based on the status of the order.

Commerce's `sendEmail()` method assumes the `Email`'s `to` property will have a value, since a validation error will occur if I try saving the email in the CP Email Settings and the "Recipient" is set to "Send to custom recipient".  But this form also accepts Twig, and I have a method on my `OrderBehavior` class I can call in that form field that will try to construct the list of email addresses from the users in the order's custom relation field I've created. It is possible in my situation, however, that there could be no users in this field that could/should receive the email, and thus an empty string gets returned to the `Email`'s `to` property. If that happens, [then an exception is thrown](https://github.com/craftcms/commerce/blob/375fd16cc1f00a85f1aea90ee5e87413f78c660c/src/services/Emails.php#L467) trying to set `to` on the mailer.

Up to this point, (and to my knowledge) my code has no way to listen for an event and stop the email from getting prepared, if such an issue occurs. The next event I can respond to is `EVENT_BEFORE_SEND_MAIL`, which happens well after all the email's to/from/cc/etc properties are set and its template is rendered. It would also be nice to be able to inject extra variables into the template before it gets parsed, or change certain variables, which this event would allow. 

I just wish my module had the option to affect the `Email` model before `sendEmail` starts checking for exceptions.

**Comment by nfourtythree** [comment, marked as answer]
Hi @modelesque 

Thank you for your message.

There are a couple of things that you can do here to get things working for you.

Firstly, in terms of the sending to the recipients, if you are using Twig to render the output instead of having it fallback to `''` you could set it to something like `donotsend@example.com`. You could then be checking for this value in the "before send email" event to prevent sending.

When it comes to adding extra variables, the body of the email is rendered like a standard template so you are able to access template-related events. You could use the ["before render template" event](https://github.com/craftcms/cms/blob/develop/src/web/View.php#L1791-L1796) to add anything extra you need.

There are certainly improvements that we can look into, to make things easier in the future. We will add the process to the list of things to review for future updates.

Thanks!

**Reply by modelesque** [reply]
Hi @nfourtythree 

All good, thanks for your response. Those are some good ideas. I may also listen for the `src/services/OrderHistories::EVENT_ORDER_STATUS_CHANGE` event and send out the emails myself so my code has complete control over the parsing of the templates.
